File: algorithms/algorithms-cplusplus/leetcode/Question_0031_Next_Permutation.cc
#include <vector>

using namespace std;

class Solution {
 public:
  void nextPermutation(vector<int>& nums) {
    if (nums.size() < 2) {
      return;
    }
    int i = nums.size() - 2;
    // 找到第一个比后一个数小的数
    while (i >= 0 && nums[i + 1] <= nums[i]) {
      i--;
    }
    if (i >= 0) {
      int j = nums.size() - 1;
      // 找到第一个大于nums[i]的数
      while (j >= 0 && nums[j] <= nums[i]) {
        j--;
      }
      swap(nums[i], nums[j]);
    }
    reverse(nums, i + 1);
  }

  void reverse(vector<int>& nums, int l) {
    int r = nums.size() - 1;
    while (l < r) {
      swap(nums[l++], nums[r--]);
    }
  }
};

File: algorithms/algorithms-python/leetcode/Question_0031_Next_Permutation.py
#!/usr/bin/python
# coding: utf-8

class Solution(object):
    def nextPermutation(self, nums):
        """
        :type nums: List[int]
        :rtype: void Do not return anything, modify nums in-place instead.
        """
        n = len(nums)
        if n < 2:
            return
        i = n - 2
        while i >= 0 and nums[i] >= nums[i + 1]:
            i -= 1
        if i >= 0:
            j = n - 1
            while j >= 0 and nums[j] <= nums[i]:
                j -= 1
            self.swap(nums, i, j)
        self.reverse(nums, i + 1)
    
    def reverse(self, nums, start):
        i, j= start, len(nums) - 1
        while i < j:
            self.swap(nums, i, j)
            i += 1
            j -= 1

    def swap(self, nums, i, j):
        tmp = nums[i]
        nums[i] = nums[j]
        nums[j] = tmp
